codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java

```diff
@@ -111,123 +112,152 @@ public ChannelFuture writeData(ChannelHandlerContext ctx, int streamId, ByteBuf
             int payloadLength = data.readableBytes() + padding + flags.getPaddingPresenceFieldLength();
             verifyPayloadLength(payloadLength);
 
-            ByteBuf out = ctx.alloc().buffer(FRAME_HEADER_LENGTH + payloadLength);
+            writeFrameHeader(ctx, payloadLength, DATA, flags, streamId, promiseAggregator.newPromise());
 
-            writeFrameHeader(out, payloadLength, DATA, flags, streamId);
-
-            writePaddingLength(padding, out);
+            writePaddingLength(ctx, padding, promiseAggregator.newPromise());
+        } catch (Throwable t) {
+            data.release();
+            return promiseAggregator.tryFailure(t);
+        }
 
+        try {
             // Write the data.
-            out.writeBytes(data, data.readerIndex(), data.readableBytes());
+            ctx.write(data, promiseAggregator.newPromise());
 
             // Write the required padding.
-            out.writeZero(padding);
-            return ctx.write(out, promise);
-        } catch (RuntimeException e) {
-            return promise.setFailure(e);
-        } finally {
-            data.release();
+            writeZero(ctx, padding, promiseAggregator.newPromise());
+            promiseAggregator.doneAllocatingPromises();
+            return promise;
+        } catch (Throwable t) {
+            return promiseAggregator.tryFailure(t);
         }
     }
 
     @Override
     public ChannelFuture writeHeaders(ChannelHandlerContext ctx, int streamId,
             Http2Headers headers, int padding, boolean endStream, ChannelPromise promise) {
-        return writeHeadersInternal(ctx, promise, streamId, headers, padding, endStream,
-                false, 0, (short) 0, false);
+        return writeHeadersInternal(ctx, streamId, headers, padding, endStream,
+                false, 0, (short) 0, false, promise);
     }
 
     @Override
     public ChannelFuture writeHeaders(ChannelHandlerContext ctx, int streamId,
             Http2Headers headers, int streamDependency, short weight, boolean exclusive,
             int padding, boolean endStream, ChannelPromise promise) {
-        return writeHeadersInternal(ctx, promise, streamId, headers, padding, endStream,
-                true, streamDependency, weight, exclusive);
+        return writeHeadersInternal(ctx, streamId, headers, padding, endStream,
+                true, streamDependency, weight, exclusive, promise);
     }
 
     @Override
     public ChannelFuture writePriority(ChannelHandlerContext ctx, int streamId,
```

Comment:
@louiscryan - Good point. Lets re-evaluate after I push my next change which optimized and reduces buffer allocations for the current interface.  This particular method is only 1 buffer allocation in the next commit.